c++/src/BpackingAvx512.cc

```diff
@@ -0,0 +1,4476 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BpackingAvx512.hh"
+#include "BitUnpackerAvx512.hh"
+#include "CpuInfoUtil.hh"
+#include "RLEv2.hh"
+
+namespace orc {
+  UnpackAvx512::UnpackAvx512(RleDecoderV2* dec) : decoder(dec), unpackDefault(UnpackDefault(dec)) {
+    // PASS
+  }
+
+  UnpackAvx512::~UnpackAvx512() {
+    // PASS
+  }
+
+  void UnpackAvx512::vectorUnpack1(int64_t* data, uint64_t offset, uint64_t len) {
+    uint32_t bitWidth = 1;
+    const uint8_t* srcPtr = reinterpret_cast<const uint8_t*>(decoder->bufferStart);
+    uint32_t numElements = 0;
+    int64_t* dstPtr = data + offset;
+    uint64_t bufMoveByteLen = 0;
+    uint64_t bufRestByteLen = decoder->bufferEnd - decoder->bufferStart;
+    bool resetBuf = false;
+    uint64_t startBit = 0;
+    uint64_t tailBitLen = 0;
+    uint32_t backupByteLen = 0;
+
+    while (len > 0) {
+      if (startBit != 0) {
+        bufMoveByteLen +=
+            moveLen(len * bitWidth + startBit - ORC_VECTOR_BYTE_WIDTH, ORC_VECTOR_BYTE_WIDTH);
+      } else {
+        bufMoveByteLen += moveLen(len * bitWidth, ORC_VECTOR_BYTE_WIDTH);
+      }
+
+      if (bufMoveByteLen <= bufRestByteLen) {
+        numElements = len;
+        resetBuf = false;
+        len -= numElements;
+      } else {
+        if (startBit != 0) {
+          numElements =
+              (bufRestByteLen * ORC_VECTOR_BYTE_WIDTH + ORC_VECTOR_BYTE_WIDTH - startBit) /
+              bitWidth;
+          len -= numElements;
+          tailBitLen = fmod(
+              bufRestByteLen * ORC_VECTOR_BYTE_WIDTH + ORC_VECTOR_BYTE_WIDTH - startBit, bitWidth);
+          resetBuf = true;
+        } else {
+          numElements = (bufRestByteLen * ORC_VECTOR_BYTE_WIDTH) / bitWidth;
+          len -= numElements;
+          tailBitLen = fmod(bufRestByteLen * ORC_VECTOR_BYTE_WIDTH, bitWidth);
+          resetBuf = true;
+        }
+      }
+
+      if (tailBitLen != 0) {
+        backupByteLen = tailBitLen / ORC_VECTOR_BYTE_WIDTH;
+        tailBitLen = 0;
+      }
+
+      if (startBit > 0) {
+        uint32_t align = getAlign(startBit, bitWidth, 8);
+        if (align > numElements) {
+          align = numElements;
+        }
+        if (align != 0) {
+          bufMoveByteLen -=
+              moveLen(align * bitWidth + startBit - ORC_VECTOR_BYTE_WIDTH, ORC_VECTOR_BYTE_WIDTH);
+          plainUnpackLongs(dstPtr, 0, align, bitWidth, startBit);
+          srcPtr = reinterpret_cast<const uint8_t*>(decoder->bufferStart);
+          bufRestByteLen = decoder->bufferEnd - decoder->bufferStart;
+          dstPtr += align;
+          numElements -= align;
+        }
+      }
```

Comment:
Codes of line 46 to 93 are similar (or identical?) in these methods, e.g. the same codes appear in vectorUnpack2 and vectorUnpack3. Can we extract them to reduce the method size?

I think the core of these methods are the while-loop of `while (numElements >= VECTOR_UNPACK_8BIT_MAX_NUM)`. It'd be nice to refactor the codes dealing with boundaries and left these methods focus on this while-loop.

Reply:
Thank you very much for reminding me. 
Already added 2 inline functions to optimizing the bit-unpacking code path. Please check it.
https://github.com/wpleonardo/orc/blob/0d59f902680ae279fbcd49e44696ad2f84e1264e/c%2B%2B/src/BpackingAvx512.hh#L74
https://github.com/wpleonardo/orc/blob/0d59f902680ae279fbcd49e44696ad2f84e1264e/c%2B%2B/src/BpackingAvx512.hh#L80